### PR TITLE
storage: move PrivateLink status updates into storage controller

### PR DESCRIPTION
**Pull request by sploiselle**
We go through great length to ensure we never went backward in time when producing VPCEndpointEvents as statuses. However, the core of the design is exactly the same as the other semantics we want for other collections managed by the `CollectionStatusManager`. Rather than having these same bits of logic in very disparate places, unify them.

### Motivation

This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-controller/Cargo.toml [hunk 1]
```diff
@@ -33,8 +34,16 @@ proptest = { version = "1.0.0", default-features = false, features = ["std"] }
 prost = { version = "0.11.3", features = ["no-recursion-limit"] }
 serde = { version = "1.0.152", features = ["derive"] }
 serde_json = { version = "1.0.89" }
-timely = { version = "0.12.0", default-features = false, features = ["bincode"] }
-tokio = { version = "1.24.2", features = ["fs", "rt", "sync", "test-util", "time"] }
+timely = { version = "0.12.0", default-features = false, features = [
```
Comment (sploiselle) [inline src/storage-controller/Cargo.toml:37]: I don't know why my autoformatter started doing this. Will see if I can fix it on next push.

File: src/storage-controller/src/lib.rs [hunk 2]
```diff
@@ -2078,6 +2108,15 @@ where
         self.append_to_managed_collection(id, updates).await;
     }
 
+    async fn append_privatelink_updates(&mut self, updates: Vec<VpcEndpointEvent>) {
```
Comment (petrosagg) [inline src/storage-controller/src/lib.rs:2111]: Does the storage controller know or care about what a privatelink connection is and the fact that it has status updates? If this is some process driven by the adapter why are we not using the generic table append API to add whatever status updates the adapter needs?

File: src/storage-controller/src/healthcheck.rs [hunk 3]
```diff
@@ -8,20 +8,63 @@
...
+    fn id(&self) -> GlobalId;
+
+    /// Provide this record's value, which is used as the "last value" when
+    /// determining whether or not to produce values.
+    fn value(self) -> String;
+}
+
+// Dependency graph means we have to implement this.
+impl IntrospectionStatusUpdate for VpcEndpointEvent {
+    fn produce_vs_last_value(&self, last_value: &str) -> bool {
```
Comment (rjobanp) [inline src/storage-controller/src/healthcheck.rs:50]: this abstraction is great!

File: src/storage-controller/src/lib.rs [hunk 4]
```diff
@@ -933,46 +929,80 @@ where
...
+                                            .expect("schema has not changed")
+                                            .unwrap_str()
+                                            .into(),
+                                    )
+                                }),
                             )
-                            .await;
                         }
                         IntrospectionType::PrivatelinkConnectionStatusHistory => {
-                            // Truncate the private link connection status history table and
```
Comment (rjobanp) [inline src/storage-controller/src/lib.rs:948]: I think it'd be helpful to leave a comment like this somewhere, explaining that for the privatelink statuses we use the `occured_at` timestamp as the comparison for whether to store new events, but for other status tables we use the `status` column value itself.

**Review by jkosh44 — APPROVED** [review]
Adapter changes LGTM

File: src/storage-controller/src/healthcheck.rs [hunk 5]
```diff
@@ -8,20 +8,63 @@
...
+        true
+    }
+
+    /// Provide this record's GlobalId, which is used as a key when determining
+    /// whether or not to produce values.
+    fn id(&self) -> GlobalId;
+
+    /// Provide this record's value, which is used as the "last value" when
+    /// determining whether or not to produce values.
+    fn value(self) -> String;
```
Comment (rjobanp) [inline src/storage-controller/src/healthcheck.rs:45]: I'm a little concerned that the `String` type here might lead to future bugs if someone is not careful when implementing this trait.. It assumes that the string values given to `self.collection_status_manager.extend_previous_statuses` for this collection match up with this string `value` here, which seems error-prone.

I'm not sure of a better way to enforce this without just storing whole rows and using the column index to ensure consistency, any ideas? If there isn't an elegant way to do this I'm fine to merge as-is, just wanted to call it out.

**Review by rjobanp — COMMENTED** [review]
thank you so much for refactoring this! Looks much cleaner than before

**Comment by sploiselle** [comment]
Closing in favor of #24137

File: src/storage-controller/src/lib.rs (hunk 2 above)
Comment (sploiselle) [reply src/storage-controller/src/lib.rs:2111]: The controller had another `...privatelink...` method, and these commits moved things in a more positive direction, so went with it. I hear you, though, and I wasn't totally satisfied with this abstraction leakage, so I refactored it to allow any introspection collection to be appended to via the public API. https://github.com/MaterializeInc/materialize/pull/24137/files#diff-3154d32c4d0818b6513328628453395000c38bd1238bd1fa9e96b796201b38a6R521

**Pull request closed without merging**